### PR TITLE
initial work on sdc-cloudapi for PUBAPI-1420

### PR DESCRIPTION
**Pull request by joshwilsdon**
initial work

### CODE REVIEW

File: lib/machines.js [hunk 1]
```diff
@@ -1113,6 +1114,93 @@ function dnsify(str) {
...
+    if (mountpoint.search(/[^\/]/) === -1) {
+        return false;
+    }
+
+    return true;
+}
+
+// stops at first error
+function validateVolume(volume) {
+    var VALID_VOLUME_MODES = ['ro', 'rw'];
```
Comment (misterdjules) [inline lib/machines.js]: Should we `assert.object(volume, 'volume');`?

File: lib/machines.js [hunk 2]
```diff
@@ -1113,6 +1114,93 @@ function dnsify(str) {
...
+
+    if (volume.type !== undefined && volume.type !== 'tritonnfs') {
+        return (new InvalidArgumentError('unknown volume type "' + volume.type
```
Comment (misterdjules) [inline lib/machines.js]: If we return on the first error, then we lose the opportunity to report all relevant errors for a given request. Should we instead return an array of errors?

File: lib/machines.js [hunk 3]
```diff
@@ -1113,6 +1114,93 @@ function dnsify(str) {
...
+function validateVolumes(volumes) {
+    var idx = 0;
+    var validationErr;
+
+    assert.arrayOfObject(volumes, 'volumes');
+
+    for (idx = 0; idx < volumes.length; idx++) {
+        validationErr = validateVolume(volumes[idx]);
+        if (validationErr) {
+            break;
```
Comment (misterdjules) [inline lib/machines.js]: Same here: we early out on the first validation error, should we instead validate all input and report all relevant errors?

File: test/volumes-automount.test.js [hunk 4]
```diff
@@ -0,0 +1,626 @@
...
+var mod_testConfig = require('./lib/config');
+var mod_testNetworks = require('./lib/networks');
+var mod_testVolumes = require('./lib/volumes');
+
+var machinesCommon = require('./machines/common');
+
+var CONFIG = mod_config.configure();
+var IMGAPI_SOURCE = 'https://images.joyent.com';
+var KEY_FILENAME = '/tmp/cloudapi-test-key';
+var TEST_IMAGE_LX = '7b5981c4-1889-11e7-b4c5-3f3bdfc9b88b'; // ubuntu-16.04
```
Comment (misterdjules) [inline test/volumes-automount.test.js]: Is it safe to hardcode it?

File: test/volumes-automount.test.js [hunk 5]
```diff
@@ -0,0 +1,626 @@
...
+var TEST_IMAGE_SMARTOS =
+    'ede31770-e19c-11e5-bb6e-3b7de3cca9ce'; // minimal-multiarch-lts (15.4.1)
```
Comment (misterdjules) [inline test/volumes-automount.test.js]: Same regarding hardcoding of the image UUID.

File: test/volumes-automount.test.js [hunk 6]
```diff
@@ -0,0 +1,626 @@
...
+var UFDS_ADMIN_UUID = CONFIG.ufds_admin_uuid;
+
+// XXX
```
Comment (misterdjules) [inline test/volumes-automount.test.js]: What does this `// XXX` mean? I'd suggest replacing that with a clearer comment.

File: test/volumes-automount.test.js [hunk 7]
```diff
@@ -0,0 +1,626 @@
...
+                    }
+                }
+
+                t.ok(networkUuidSsh, 'found external network: '
+                    + networkUuidSsh);
+                t.end();
+            });
+    });
+
+    // need a package we can use to provision our containers
```
Comment (misterdjules) [inline test/volumes-automount.test.js]: Should that comment be about networks instead of packages?

File: test/volumes-automount.test.js [hunk 8]
```diff
@@ -0,0 +1,626 @@
...
+        var payload;
+
+        payload = {
+            metadata: {},
+            image: TEST_IMAGE_LX,
+            package: testPackage.id,
+            name: 'cloudapi-volume-lx-' + libuuid.create().split('-')[0],
+            firewall_enabled: false,
+            networks: [
+                {ipv4_uuid: networkUuidSsh, primary: true},
```
Comment (misterdjules) [inline test/volumes-automount.test.js]: Nice!

File: lib/machines.js [hunk 9]
```diff
@@ -1164,6 +1253,21 @@ function create(req, res, next) {
         }
     }
 
+    if (req.params && req.params.volumes && req.params.volumes.length > 0) {
+        if (opts.brand === 'kvm') {
+            // At this time we do not support NFS volumes with KVM
+            return next(new InvalidArgumentError('volumes not yet supported ' +
```
Comment (misterdjules) [inline lib/machines.js]: Should we add a test that checks that the relevant error is returned when trying to provision a KVM VM with volumes? That can be done in a subsequent PR, but we'd need to track that with a separate ticket.

File: lib/machines.js [hunk 10]
```diff
@@ -1113,6 +1114,93 @@ function dnsify(str) {
     return (str.toLowerCase().replace(/[^a-z0-9-]+/g, '-'));
 }
 
+// blatently stolen from sdc-vmapi lib/common/validation.js
+
+function validVolumeName(volumeName) {
+    var VALID_VOLUME_NAME_REGEXP = /^[a-zA-Z0-9][a-zA-Z0-9_\.\-]+$/;
+    if (volumeName === undefined || volumeName === null ||
```
[the remark this answers is not in the document]
Comment (joshwilsdon) [reply lib/machines.js]: Sure, this makes sense.

File: lib/machines.js (hunk 1 above)
Comment (joshwilsdon) [reply lib/machines.js]: yep!

File: lib/machines.js (hunk 2 above)
Comment (joshwilsdon) [reply lib/machines.js]: Sure, I'll change to do that.

File: lib/machines.js (hunk 3 above)
Comment (joshwilsdon) [reply lib/machines.js]: Yep.

File: lib/machines.js (hunk 9 above)
Comment (joshwilsdon) [reply lib/machines.js]: I'll add a test that does this. Sure.

File: test/volumes-automount.test.js (hunk 4 above)
Comment (joshwilsdon) [reply test/volumes-automount.test.js]: I think so for now, but I'll create a ticket to revisit this in the future.

File: test/volumes-automount.test.js (hunk 5 above)
Comment (joshwilsdon) [reply test/volumes-automount.test.js]: I created https://devhub.joyent.com/jira/browse/PUBAPI-1436 for this.

File: test/volumes-automount.test.js (hunk 6 above)
Comment (joshwilsdon) [reply test/volumes-automount.test.js]: This was actually left in by mistake. I'll remove it.

File: test/volumes-automount.test.js (hunk 7 above)
Comment (joshwilsdon) [reply test/volumes-automount.test.js]: Indeed it should!

File: test/volumes-automount.test.js (hunk 8 above)
Comment (joshwilsdon) [reply test/volumes-automount.test.js]: Thanks! :)

**Comment by misterdjules** [comment]
@joshwilsdon Latest updates look good to me!

**Comment by joshwilsdon** [comment]
The "experimental_cloudapi_automount" flag stuff was pointed out to me by @misterdjules in 1:1 chat. At this point I'm considering this complete and just waiting on a final go-ahead.

**Comment by joshwilsdon** [comment]
integrated via c70cbc5ce7c62ad7beeb26cc0dba2f429931a598

**Pull request closed without merging**